### PR TITLE
[py] overload log_path to accept kwargs for open

### LINKED ISSUE #10703: [🐛 Bug]: PySelenium doesn't support a named pipe as a log file
### What happened?

As part of [OpenWPM](https://github.com/openwpm/OpenWPM) we use Selenium to drive multiple Firefox browsers in parallel.

To capture, enrich and redirect the browser's logging output, we use named pipes that look like normal log files to Selenium.
However pipes don't support seeking so the need to be opened with `w` and not with `a+` as the [`firefox.Service` currently does](https://github.com/SeleniumHQ/selenium/blob/268161f84fb63e3e3e3177eb8b99b812ee3ae413/py/selenium/webdriver/firefox/service.py#L49). [We have previously monkeypatched this](https://github.com/openwpm/OpenWPM/blob/8464741bbfcdbd9aad54686d3f290a060c168488/openwpm/deploy_browsers/selenium_firefox.py#L79-L122) but that approach has led to us breaking every time a new Selenium release comes out. [A newer approach that tries to call `firefox.Service.__init__` and open the `log_file` afterwards](https://github.com/openwpm/OpenWPM/blob/45ffd5a752e88c24607776e62cd2d53c965effb4/openwpm/deploy_browsers/selenium_firefox.py#L80-L105) doesn't work at all, so I wanted to check if there was an appetite to address this issue upstream.

### How can we reproduce the issue?

```shell
import os
from selenium.webdriver import Firefox
from selenium.webdriver.firefox.service import Service
pipe_name = "selenium_pipe.log"
os.mkfifo(pipe_name, 0o600)
service=Service(executable_path="firefox",log_path=pipe_name)
driver = Firefox(
    service=service
)
driver.get("https://example.com")
```

### Relevant log output

```shell
File "/home/vringar/selenium_mvp/demo.py", line 9, in <module>
    service=Service(executable_path="firefox",log_path=pipe_name)
File "/home/stefan/.conda/envs/openwpm/lib/python3.10/site-packages/selenium/webdriver/firefox/service.py", line 50, in __init__
    log_file = open(log_path, "a+") if log_path else None
io.UnsupportedOperation: File or stream is not seekable.
```

### Operating System

Linux

### Selenium version

Python 4.1.0

### What are the browser(s) and version(s) where you see this issue?

Firefox 100

### What are the browser driver(s) and version(s) where you see this issue?

GeckoDriver 0.30.0

### Are you using Selenium Grid?

_No response_

### PR DESCRIPTION
**Pull request by vringar**
### Description

This change allows users to exactly specify how they want
to open their log file while falling back to the defaults for arguments that weren't specified.

### Motivation and Context

In [OpenWPM](https://github.com/openwpm/OpenWPM) we run multiple Firefox instances in parallel.
To integrate their output with our logging infrastructure we want them to write their logs into pipes, which we then read on the other side, enrich it with metadata and log it ourselves.

This feature is relevant to anyone who wants to interleave their own code's log lines with Selenium's or the browser's.

Fixes #10703

### Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by vringar** [comment]
This branch is still lacking test cases, which I'll be adding once there is interest to merge. If https://github.com/SeleniumHQ/selenium/issues/10703 gets closed as invalid/not relevant I don't want to have invested too much effort.

File: py/selenium/webdriver/firefox/service.py [hunk 1]
```diff
@@ -46,7 +47,14 @@ def __init__(self, executable_path: str = DEFAULT_EXECUTABLE_PATH,
             in the services' environment.
 
         """
-        log_file = open(log_path, "a+") if log_path else None
+        log_file = None
```
Comment (symonk) [inline py/selenium/webdriver/firefox/service.py:51]: Seems good to me; we could also allow `log_path` to become an (optional) length-two tuple that accepts something like `log_path: Union[str, Tuple[str, str]]` where the later case would be the `path, mode` passed to `open(...)`.  I'm ok with the current changes tho; seems to keep the flow but support an edge case.  Could we add a test for this?

Comment (vringar) [reply py/selenium/webdriver/firefox/service.py:51]: I like this solution a lot more so I changed it to that. I hope the change to the doc string is understandable.
I'm currently unable to run the tests locally but I'll try to write one anyway.

**Comment by vringar** [comment]
I think this PR is ready for another round of review.
The test I've added is pretty basic but idk if you wanted more @symonk.

File: py/test/selenium/webdriver/marionette/mn_launcher_tests.py [hunk 2]
```diff
@@ -32,10 +33,17 @@ def test_we_can_launch_multiple_firefox_instances(capabilities):
...
 
 def test_launch_firefox_with_empty_string_service_log_path(capabilities):
-    driver = Firefox(capabilities=capabilities, service_log_path="")
+    service = Service(log_path="")
+    driver = Firefox(capabilities=capabilities, service=service)
+    driver.quit()
+
+def test_launch_firefox_with_empty_string_service_log_path(capabilities):
+    service = Service(log_path=("geckodriver.log", "w"))
+    driver = Firefox(capabilities=capabilities, service=service)
```
Comment (omkarkhatavkar) [inline py/test/selenium/webdriver/marionette/mn_launcher_tests.py]: You can parametrize tests like this?

```python
@pytest.mark.parametrize('log_path', [None, "", ("geckodriver.log", "w")])
def test_launch_firefox_with_service_log_path(log_path):
    service = Service(log_path=log_path)
    driver = Firefox(capabilities=capabilities, service=service)
    driver.quit()
```

Comment (vringar) [reply py/test/selenium/webdriver/marionette/mn_launcher_tests.py]: Good idea! Done

File: py/test/selenium/webdriver/marionette/mn_launcher_tests.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 # under the License.
 
 from selenium.webdriver import Firefox
+from selenium.webdriver.firefox.service import Service
```
Comment (omkarkhatavkar) [inline py/test/selenium/webdriver/marionette/mn_launcher_tests.py:19]: `import pytest` is missing i think

Comment (vringar) [reply py/test/selenium/webdriver/marionette/mn_launcher_tests.py:19]: True. Done.
wish I had some way to run the tests so I could see if I missed anything else.

File: py/selenium/webdriver/firefox/service.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ class Service(service.Service):
 
     def __init__(self, executable_path: str = DEFAULT_EXECUTABLE_PATH,
                  port: int = 0, service_args: List[str] = None,
-                 log_path: str = "geckodriver.log", env: dict = None):
+                 log_path: Union[str, Tuple[str, str]] = "geckodriver.log", env: dict = None):
```
Comment (symonk) [inline py/selenium/webdriver/firefox/service.py]: `Sequence[str, str]` is probably preferred here; a list would be fine etc, let's not rely on concrete implementations when we can avoid it (I know this statement is grossly untrue for other parts but i'd like to keep anything new as loose as we can get away with, checking the `isinstance` will need an update too

File: py/selenium/webdriver/firefox/service.py [hunk 5]
```diff
@@ -41,12 +41,13 @@ def __init__(self, executable_path: str = DEFAULT_EXECUTABLE_PATH,
...
         :param log_path: Optional path for the GeckoDriver to log to.
-            Defaults to _geckodriver.log_ in the current working directory.
+            Can also be called with (path_name, mode).
+            Defaults to _geckodriver.log_ in the current working directory and _a+_ respectively.
         :param env: Optional dictionary of output variables to expose
             in the services' environment.
-
         """
-        log_file = open(log_path, "a+") if log_path else None
+        log_path, log_path_mode = (log_path[0], log_path[1]) if isinstance(log_path,Tuple) else (log_path, "a+")
```
Comment (symonk) [inline py/selenium/webdriver/firefox/service.py]: How about:

```
# if log_path was not explicitly set to None
open_args = (log_path,) if isinstance(log_path, str) else log_path
log_file = open(*open_args)
```
Has the added benefit of allowing the caller to shovel arguments into the `open` call positionally (should there be another use case to change something else and not just `mode`)

(would need to account for the `Optional` nature too; docs point at optional, type hint does not).  I suspect it might be easier to default it to `None` and then assign it to `geckodriver.log` if it `is None` (I assume later calling code is _assuming_ the log file is always available).

edit2: best to keep the current behaviour if `log_path` is explicitly set to `None` after looking at some parent code

File: py/selenium/webdriver/firefox/service.py (hunk 4 above)
Comment (vringar) [reply py/selenium/webdriver/firefox/service.py]: Done!

File: py/selenium/webdriver/firefox/service.py (hunk 5 above)
Comment (vringar) [reply py/selenium/webdriver/firefox/service.py]: Okay, I think I did what you asked, but I'm not sure I 100% I understood you.

File: py/selenium/webdriver/firefox/service.py [hunk 6]
```diff
@@ -41,12 +41,15 @@ def __init__(self, executable_path: str = DEFAULT_EXECUTABLE_PATH,
...
         :param log_path: Optional path for the GeckoDriver to log to.
-            Defaults to _geckodriver.log_ in the current working directory.
+            Can also be called with (path_name, mode).
+            Defaults to _geckodriver.log_ in the current working directory and _a+_ respectively.
         :param env: Optional dictionary of output variables to expose
             in the services' environment.
-
         """
-        log_file = open(log_path, "a+") if log_path else None
+        log_file = None
```
Comment (symonk) [inline py/selenium/webdriver/firefox/service.py:51]: ignore my other comments, misread this var name, LGTM :+1:

File: py/test/selenium/webdriver/marionette/mn_launcher_tests.py [hunk 7]
```diff
@@ -31,11 +33,8 @@ def test_we_can_launch_multiple_firefox_instances(capabilities):
...
-def test_launch_firefox_with_none_service_log_path(capabilities):
-    driver = Firefox(capabilities=capabilities, service_log_path=None)
-    driver.quit()
-
-
-def test_launch_firefox_with_empty_string_service_log_path(capabilities):
-    driver = Firefox(capabilities=capabilities, service_log_path="")
-    driver.quit()
+@pytest.mark.parametrize('log_path', [None, "", ("geckodriver.log", "w")])
+def test_launch_firefox_with_service_log_path(log_path):
```
Comment (symonk) [inline py/test/selenium/webdriver/marionette/mn_launcher_tests.py:35]: can we inspect the underlying `log_file` attr for assertion? (calling `.quit()` is unnecessary also).  `driver.service.log_file` will be available;

I'd like to see:

 - not passed (`geckodriver.log`)
 - `None` explicitly
 - other file + mode

both `name` & `mode` will be available on the `log_file` attr

**Comment by symonk** [comment]
Apologies for the delays! I've approved the PR for running actions workflows.  Thanks a bunch for the pull request and taking the effort to contribute, it is greatly appreciated!  Few comments around tests

File: py/test/selenium/webdriver/marionette/mn_launcher_tests.py (hunk 7 above)
Comment (vringar) [reply py/test/selenium/webdriver/marionette/mn_launcher_tests.py:35]: I couldn't figure out how to test for the default behaviour in the parametrized version, so I split that off into an extra test. I hope that's okay.

**Comment by vringar** [comment]
No worries about the delays!
Unfortunately your permission to run the tests only seems to apply to a single commit, so I've activated the Python Actions for my fork, to make sure the tests are passing.

**Comment by vringar** [comment]
Okay, the chrome tests are failing (in the acitons of my fork) but as I didn't modify any code on that path, I don't think that should block merging here.

**Comment by vringar** [comment]
Just bumping to make sure it's not lost.

**Comment by vringar** [comment]
Hey,
is there something I can do to move this forward?

**Comment by vringar** [comment]
@symonk Sorry to ping you on this but can you run the tests again please

**Comment by titusfortner** [comment]
It doesn't make sense to me for a parameter named `log_path` to take a dictionary that takes 3 optional arguments. Does Python Selenium manage logging this way elsewhere in the code or is it only geckodriver logs? Would it make sense to create a new parameter/class? (that's what Java would do, I'm not actually sure what the Python idiomatic approach would be)

I'm also thinking along the lines of making all logging more easy and have one class to manage all the sources and outputs of logs. #10944

**Comment by vringar** [comment]
> Does Python Selenium manage logging this way elsewhere in the code or is it only geckodriver logs?

I found the [`webkitgtk/service.py`](https://github.com/SeleniumHQ/selenium/blob/316f9738a8e2079265a0691954ca8847e68c598d/py/selenium/webdriver/webkitgtk/service.py#L39) and [`wpewebkit/service.py`](https://github.com/SeleniumHQ/selenium/blob/316f9738a8e2079265a0691954ca8847e68c598d/py/selenium/webdriver/wpewebkit/service.py#L39) have a similar pattern, but they use binary logs and always use `wb` open mode, so they are already different from the Firefox service. Chromium browsers just pass the log location as an Command line argument.

> Would it make sense to create a new parameter/class? (that's what Java would do, I'm not actually sure what the Python idiomatic approach would be)

AFAIK, having a parameter accept multiple types is the usual way to do overloading, as there is no overload resolution in Python and always only a single constructor. I agree the name is unfortunate, but changing it would be backwards incompatible.

> It doesn't make sense to me for a parameter named log_path to take a dictionary that takes 3 optional arguments.

As a technicality, the dict can have [up to 8 values](https://docs.python.org/3/library/functions.html#open) since that is how many the `open` function accepts. I could also break out the `encoding` and `mode` as separate optional params, which is what I initially did until @symonk [guided me](https://github.com/SeleniumHQ/selenium/pull/10704#discussion_r883936345) to the overload approach.
